src/Uno.UI/UI/Xaml/Controls/ToolTip/ToolTip.cs

```diff
@@ -73,10 +129,600 @@ private void AttachToPopup()
+					top = maxY - toolTipRect.Height - TOOLTIP_TOLERANCE;
+				}
+
+				if (top < 0)
+				{
+					// If the upper edge of Plug-in obscures the ToolTip,
+					// the control repositions itself to align with the upper edge.
+					// align with the top of the plug-in
+					top = 0;
+				}
+
+				if (left + toolTipRect.Width > maxX)
+				{
+					// If the right edge obscures the ToolTip,
+					// it opens in the opposite direction from the obscuring edge.
+					left = maxX - toolTipRect.Width - TOOLTIP_TOLERANCE;
+				}
+
+				if (left < 0)
+				{
+					// If the left edge obscures the ToolTip,
+					// it then aligns with the obscuring screen edge
+					left = 0;
+				}
+
+				// if right/bottom doesn't fit into the plug-in bounds, clip the ToolTip
+				{
+					var clipCalculationsRect = new Rect(left, top, toolTipRect.Width, toolTipRect.Height);
+					CalculateTooltipClip(clipCalculationsRect, maxX, maxY);
+				}
+
+				// The placement algorithm operates in LTR mode (with transformed data if it
+				// is really in RTL mode), so we also need to transform the X value it returns.
+
+				if (bIsRTL)
+				{
+					left = screenWidth - left;
+				}
+
+				// position the parent Popup
+				spPopup.VerticalOffset = top + verticalOffset;
+				spPopup.HorizontalOffset = left + horizontalOffset;
+			}
+		}
+
+		// Sets the location of the ToolTip's Popup within the Xaml window.
+		private void PerformNonMousePlacementWithPopup(
+			Rect? pTargetRect, Rect pDimentions, PlacementMode placement)
+		{
+
+			var tooltipActualWidth = pDimentions.Width;
+			var tooltipActualHeight = pDimentions.Height;
+			var horizontalOffset = pDimentions.Left;
+			var verticalOffset = pDimentions.Top;
+
+			var bIsRTL = false;
+
+			var origin = default(Point);
+			var rcDockTo = default(Rect);
+			var szFlyout = default(Size);
+
+			var spTarget = _owner as FrameworkElement;
+			_Debug.Assert(spTarget != null, "pTarget expected to be non-null in ToolTip_Partial::PerformPlacement()");
+			if (spTarget != null)
+			{
+				// UNO TODO
+				//	        xaml::FlowDirection targetFlowDirection = xaml::FlowDirection_LeftToRight;
+
+				//spTarget->get_FlowDirection(&targetFlowDirection);
+				//bIsRTL = targetFlowDirection == xaml::FlowDirection_RightToLeft;
+
+				// We should not do placement if the target is no longer in the live tree.
+				if (!spTarget.IsLoaded)
+				{
+					return;
+				}
+			}
+
+			var spPopup = _popup;
+			if (spPopup == null)
+			{
+				return;
+			}
+
+			// For ToolTips opened by keyboard focus, PlacementMode_Mouse doesn't make any sense.
+			// Fall back to the default - PlacementMode_Top.
+			if (placement == PlacementMode.Mouse)
+			{
+				placement = PlacementMode.Top;
+			}
+
+			if (spTarget == null)
+			{
+				return;
+			}
+
+			var visibleRect = Windows.UI.Xaml.Window.Current.Bounds;
+			var constraint = visibleRect;
+
+			var windowRect = Windows.UI.Xaml.Window.Current.Bounds;
+			origin.X = windowRect.X;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['origin' is null on at least one execution path.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5404644)